### ISSUE TITLE
`Invoke-WebRequest` and `Invoke-RestMethod`: display formatting of error responses is verbose and multiline with the default, single-line error view

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#24350]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

```powershell
# Ditto for Invoke-RestMethod
Invoke-WebRequest https://example.org/NoSuchPageExists
```

### Expected behavior

Something along the lines of:

```console
Invoke-WebRequest: 404 - Not Found (https://example.org/NoSuchPageExists)
```

### Actual behavior

_Update_: I don't see this specific symptom anymore, but I now see verbose, _multiline_ error messages including the HTML source code of the 404 page; with the default `$ErrorView` in effect, `ConciseView`, a _single-line_ error message should print, such as the one suggested above.

```console
Invoke-WebRequest:



		404 - Not Found


		404 - Not Found
```

### Environment data

```powershell
PowerShell 7.5.0-preview.3
```

**Comment by TobiasPSP** [comment]
I can repro in WPS, seems to be a long standing bug. Unless I am completely mistaken, when the page does not exist, the content is bypassed directly onto the console. No way of redirecting it to a variable, not even by using `*>&1`.

**Comment by mklement0** [comment]
Hmm... now I cannot reproduce the symptom at all anymore, and instead get a verbose multiline error message that includes the HTML source code of the 404 page.

That said, with the default `$ErrorView`, `Concise`, the error message should still be _single-line_.

The reason that redirections don't work is that the web cmdlets emit (statement-)_terminating_ errors (`-ErrorVariable` works, though).

**Comment by TobiasPSP** [comment]
You have to use `-SkipHttpErrorCheck` in order to capture the data in a variable or stream. Apparently, that's the design from early days.

**Comment by mklement0** [comment]
Yes, you can (in PS 7 only) but that's a separate issue (you can alternatively get the response from the error record's `.InnerException.Response` property).

The issue at hand is that the formatting of the error message should be single-line.

**Comment by schubfre** [comment]
Is this the same problem or another one?
(I can't really show anything, because this is an internal REST API and I have no idea where i could find a true nested JSON  [a JSON within in a JSON] on the web to reproduce this freely)

PS 7.3.8 (correct behaviour)
```Powershell
Invoke-WebRequest: {"errorMessage":"{\"requestId\":\"REDACTED\",\"httpStatus\":{\"statusCode\":404,\"reasonPhrase\":\"Not Found\"},\"errorCode\":\"NOT_FOUND_ERROR\",\"errorMessage\":\"REDACTED\",\"errorSlots\":{\"1\":\"REDACTED\",\"0\":\"\"}}","httpStatus":404}
```

PS 7.4.5 (bugged, incorrect behaviour)
```Powershell
Invoke-WebRequest:
{
  "errorMessage": "{\u0022requestId\u0022:\u0022REDACTED\u0022,\u0022httpStatus\u0022:{\u0022statusCode\u0022:404,\u0022reasonPhrase\u0022:\u0022Not Found\u0022},\u0022errorCode\u0022:\u0022NOT_FOUND_ERROR\u0022,\u0022errorMessage\u0022:\u0022REDACTED\u0022,\u0022errorSlots\u0022:{\u00221\u0022:\u0022REDACTED\u0022,\u00220\u0022:\u0022\u0022}}",
  "httpStatus": 404
}
```

Firstly, you can see the single-line is gone in the second example, although the API sent out the response on a single line **without any doubt**
Secondly, you can see every, once backslash-escaped, double quote of the inner JSON body is now html(?) encoded. This incorrect formatting happens at client level (correct in PS 7.3.x, incorrect in PS 7.4.x) and not server level

If I am right, this bug was introduced in 7.4.0
And there is not a single property of the error object, which stores the correct unformatted response of the API

You can reproduce this by having a webserver answer a POST request with any "error-level" http statuscode (e.g. 404, 400, etc) and the response body having a JSON within a JSON. You can do this by introducing a JSON String into a hash and then feeding that into Convertto-JSON. The inner JSON gets escaped, while the outer JSON is a normal JSON.

I am unsure if I should create a new bug report for this or if this issue here is the same, please advise

**Comment by vexx32** [comment]
I have some still pretty scuffed work towards this here: https://github.com/vexx32/PowerShell/tree/webcmdlets-errorview

I think as a minimum we should still add:

- A `Content` property to the exception that is thrown and ensure this error handling populates it with the _raw_ content string, not least because it is a _pain_ to get that out of the inner exception in practice.
- Maybe it's worth having a `JsonResponse` and `XmlResponse` discrete properties to attach to the exception as well? The quick and dirty as I did here is to have a `object` property that either could be attached to if they parse as JSON/XML and _giving the user the structured object in the first place_, which IMO is not terrible since handling it in PowerShell will often be dynamically type checking things anyway, but it might be nice to have it as separate properties.
- Figure out whether it's worth actually printing the content of the response as we do; I suspect if we hand back the object(s) generated, we don't really _need_ to,  users can get it out of the exception directly if they want it. But I have code that will at least generate the single-line error to start, and the rest I can look at a bit later.

**Comment by schubfre** [comment]
> Is this the same problem or another one? (I can't really show anything, because this is an internal REST API and I have no idea where i could find a true nested JSON [a JSON within in a JSON] on the web to reproduce this freely)
> 
> PS 7.3.8 (correct behaviour)
> 
> Invoke-WebRequest: {"errorMessage":"{\"requestId\":\"REDACTED\",\"httpStatus\":{\"statusCode\":404,\"reasonPhrase\":\"Not Found\"},\"errorCode\":\"NOT_FOUND_ERROR\",\"errorMessage\":\"REDACTED\",\"errorSlots\":{\"1\":\"REDACTED\",\"0\":\"\"}}","httpStatus":404}
> PS 7.4.5 (bugged, incorrect behaviour)
> 
> Invoke-WebRequest:
> {
>   "errorMessage": "{\u0022requestId\u0022:\u0022REDACTED\u0022,\u0022httpStatus\u0022:{\u0022statusCode\u0022:404,\u0022reasonPhrase\u0022:\u0022Not Found\u0022},\u0022errorCode\u0022:\u0022NOT_FOUND_ERROR\u0022,\u0022errorMessage\u0022:\u0022REDACTED\u0022,\u0022errorSlots\u0022:{\u00221\u0022:\u0022REDACTED\u0022,\u00220\u0022:\u0022\u0022}}",
>   "httpStatus": 404
> }
> Firstly, you can see the single-line is gone in the second example, although the API sent out the response on a single line **without any doubt** Secondly, you can see every, once backslash-escaped, double quote of the inner JSON body is now html(?) encoded. This incorrect formatting happens at client level (correct in PS 7.3.x, incorrect in PS 7.4.x) and not server level
> 
> If I am right, this bug was introduced in 7.4.0 And there is not a single property of the error object, which stores the correct unformatted response of the API
> 
> You can reproduce this by having a webserver answer a POST request with any "error-level" http statuscode (e.g. 404, 400, etc) and the response body having a JSON within a JSON. You can do this by introducing a JSON String into a hash and then feeding that into Convertto-JSON. The inner JSON gets escaped, while the outer JSON is a normal JSON.
> 
> I am unsure if I should create a new bug report for this or if this issue here is the same, please advise
> 
> EDIT: This may have been the commit, that broke the error message [359f5c3](https://github.com/PowerShell/PowerShell/commit/359f5c35f1f4c8715da5203956858198e2b1ab98)

Is there any news on this? I still don't know if my case is related to this issue or if I should create a new issue.

**Comment by schubfre** [comment]
Is this still being worked on?

**Comment by mklement0** [comment]
@schubfre: The issue is labeled Up-for-Grabs, which implies that while the core team is unlikely to do the work themselves, community PRs are welcome.

**Comment by schubfre** [comment]
How was this completed? I don't see a linked pull request.
Does this bot just close issues without a resolution?
